callback_test.go

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023 The Ebitengine Authors
+
+//go:build darwin
+
+package purego_test
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"testing"
+	"unsafe"
+
+	"github.com/ebitengine/purego"
+)
+
+// TestCallGoFromSharedLib is a test that checks for stack corruption on darwin/arm64
+// when C calls Go code from a non-Go thread in a dynamically loaded share library.
+func TestCallGoFromSharedLib(t *testing.T) {
+	libFileName := filepath.Join(t.TempDir(), "libcbtest.so")
+	t.Logf("Build %v", libFileName)
+
+	if err := buildSharedLib(libFileName, filepath.Join("libcbtest", "callback.c")); err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(libFileName)
+
+	lib, err := purego.Dlopen(libFileName, purego.RTLD_NOW|purego.RTLD_GLOBAL)
+	if err != nil {
+		t.Fatalf("Dlopen(%q) faled: %v", libFileName, err)
+	}
+
+	var initLib func()
+	purego.RegisterLibFunc(&initLib, lib, "initLib")
+	initLib()
+
+	var printString func(p uintptr, s string)
+	purego.RegisterLibFunc(&printString, lib, "printString")
+
+	var wg sync.WaitGroup
+
+	goFunc := func(cstr *byte, n int) int {
+		s := string(unsafe.Slice(cstr, n))
+		t.Logf("FROM Go: %s\n", s)
+		wg.Done()
+		return 1
+	}
+
+	cb := purego.NewCallback(goFunc)
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		printString(cb, "a test string")
+	}
+
+	wg.Wait()
+}
+
+func buildSharedLib(libFile string, sources ...string) error {
+	out, err := exec.Command("go", "env", "CC").Output()
+	if err != nil {
+		return fmt.Errorf("go env error: %w", err)
+	}
+
+	cc := strings.TrimSpace(string(out))
+	if cc == "" {
+		return errors.New("CC not found")
+	}
+
+	args := append([]string{"-shared", "-Wall", "-Werror", "-o", libFile}, sources...)
+	cmd := exec.Command(cc, args...)
+	if _, err := cmd.Output(); err != nil {
```

Comment:
This ignores stderr. CombinedOutput seems fine

Reply:
Done. Also log the output.